man/fbtstmac.8
.TH FBTSTMAC 8 "12 September 1995"
.FB
.de SF
.nf
.nj
.sp 1
.ft CW
.in +3
..
.de UF
.XF
.sp 1
..
.de XF
.ft R
.ju
.fi
.in -3
..
.SH NAME
fbtstmac \- interactive tool to validate syntax of \fIFirstBase\fP
macro files
.SH SYNOPSIS
.B fbtstmac
.SH DESCRIPTION
.I Fbtstmac
is a simple, interactive tool that allows the user to submit macro code to
the macro parser to ensure the code is syntactically correct.
.PP
When invoked, \fIfbtstmac\fP responds with a \fIfbtestmacro>\fP prompt.
The following is a list of commands that can be used at this point:
.SF
help           - general help
exit           - exit 
quit           - exit 
^D (EOF)       - exit 
cd             - change directory
pwd            - print working directory
ls             - list working directory
!command       - submit command to UNIX shell
filename       - submit file to macro parser
.UF
When a macro file parses correctly, the message "macro file parsed AOK" will
be returned. Otherwise, a line numbered listing of the macro file and a
display of the errors detected is shown.
.SH SEE ALSO
dbedit(1), dbmacro(1), macro(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/fbtstmac.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: fbtstmac.c,v 9.1 2001/01/16 02:46:48 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Cdbtestmacro_sid[] = "@(#) $Id: fbtstmac.c,v 9.1 2001/01/16 02:46:48 john Exp $";
#endif

#include <fb.h>
#include <fb_vars.h>
#include <macro_v.h>
#include <pwd.h>

char macrotrace_filename[FB_MAXNAME];

static char user_home[FB_MAXNAME];

static pwd_home();
static help();

void change_dir();

/*
 * fbtestmacro - provide simple interface for parsing macros.
 */

   main()

      {

         char line[FB_MAXLINE], tname[FB_MAXNAME], command[FB_MAXLINE];
         int j, traceflag = -1, i;
         char *ep, buf[FB_MAXLINE];

         cdb_batchmode = 1;
         strcpy(tname, "/usr/tmp/fbXXXXXX");
         close(mkstemp(tname));
         strcpy(macrotrace_filename, tname);
         user_home[0] = NULL;
         if ((ep = getenv("HOME")) != 0){
            strcpy(user_home, ep);
            fb_assure_slash(user_home);
            }

         for (;;){
            printf("fbtestmacro> ");
            fflush(stdout);
            if (fgets(line, FB_MAXLINE, stdin) == NULL){
               printf("\n");
               break;
               }
            if (equal(line, "\n"))
               continue;
            else if (strncmp(line, "quit", 4) == 0
                  || strncmp(line, "exit", 4) == 0)
               break;
            else if (strncmp(line, "trace", 5) == 0){
               traceflag = -traceflag;
               continue;
               }
            else if (strncmp(line, "cd", 2) == 0){
               change_dir(line + 3);
               continue;
               }
            else if (strncmp(line, "pwd", 3) == 0){
               fb_getwd(buf);
               printf("%s\n", buf);
               continue;
               }
            else if (strncmp(line, "ls\n", 3) == 0){
               fb_system("ls", FB_NOROOT);
               continue;
               }
            else if (strncmp(line, "help", 4) == 0){
               help();
               continue;
               }
            else if (line[0] == '!'){
               fb_system(line + 1, FB_NOROOT);
               continue;
               }
            j = strlen(line);
            if (line[j-1] == '\n')
               line[j-1] = 0;
            if (access(line, 0) != 0){
               printf("could not find file: %s\n", line);
               continue;
               }
            symtab = NULL;
            g_symtab = NULL;
            n_ghead = NULL;
            c_ghead = NULL;
            winner = NULL;
            if (fb_macrotree(line) == FB_AOK){
               printf("%s: macro file parsed AOK\n", line);
               if (traceflag > 0){
                  printf("begin macrotrace ...\n");
                  fb_macrotrace(winner, line);
                  if ((ep = getenv("PAGER")) != 0)
                     sprintf(command, "%s %s", ep, tname);
                  else
                     sprintf(command, "more %s", tname);
                  system(command);
                  unlink(tname);
                  }
               }
            else{
               fb_tracesource(stdout, line);
               printf("%s: macro file DID NOT parse\n", line);
               }
#if X_DEBUG
            mem_check("pre garbage");
#endif
            fb_gcollect(n_ghead, c_ghead);
            fb_expunge_symtab(symtab);
            fb_expunge_symtab(g_symtab);
#if X_DEBUG
            mem_check("post garbage");
#endif
            }
      }

/*
 * change_dir - implement the change directory command 
 */

   void change_dir(dname)
      char *dname;

      {
         char buf[FB_MAXLINE], *p, pname[FB_MAXNAME], pfile[FB_MAXNAME];

         strcpy(buf, dname);
         remove_newline(buf);
         fb_rmlead(buf);
         if (buf[0] == NULL)
            strcpy(buf, user_home);
         /* do cshell style processing of ~ */
         if (buf[0] == '~'){
            if (buf[1] == '/' || buf[1] == NULL){
               strcpy(dname, user_home);
               strcat(dname, buf + 2);
               }
            else{
               /* must be ~user/xxx which means passwd lookup */
               p = strchr(buf, '/');
               if (p != 0)
                  *p = NULL;
               strcpy(pname, buf + 1);
               pfile[0] = NULL;
               if (p != 0)
                  strcpy(pfile, p + 1);
               if (pwd_home(buf, pname) == FB_ERROR){
                  fb_serror(FB_MESSAGE, "No user named", pname);
                  return;
                  }
               sprintf(dname, "%s/%s", buf, pfile);
               }
            strcpy(buf, dname);
            }
         fb_trim(buf);
         if (chdir(buf) < 0)
            fb_serror(FB_MESSAGE, "Could not change to directory", buf);
      }

/*
*  pwd_home - look up pname in the passwd file and return its home in buf
*		return FB_ERROR if not a valid pname.
*/

   static pwd_home(buf, pname)
      char *buf, *pname;

      {
         struct passwd *getpwnam(), *gp;

         gp = getpwnam(pname);
         if (gp == NULL)
            return(FB_ERROR);
         strcpy(buf, gp->pw_dir);
         return(FB_AOK);
      }

   static help()
      {
         printf( "fbtestmacro commands:\n");
         printf( "   help           - general help\n");
         printf( "   exit           - exit \n");
         printf( "   quit           - exit \n");
         printf( "   cd             - change directory\n");
         printf( "   pwd            - print working directory\n");
         printf( "   ls             - list working directory\n");
         printf( "   !command       - submit command to UNIX shell\n");
         printf( "   filename       - submit file to macro parser\n");
      }

/*
 * remove_newline - remove trailing FB_NEWLINE
 */

   remove_newline(buf)
      char *buf;

      {
         int lc;

         lc = strlen(buf) - 1;
         if (buf[lc] == FB_NEWLINE)
            buf[lc] = NULL;
      }

#if X_DEBUG
/*
 * mem_check - verify the heap of storage
 */

   mem_check(s)
      char *s;

      {
         fprintf(stderr, "Mem check @ %s: ", s);
         if (malloc_verify() == 1)
            fprintf(stderr, "PASS\n");
         else
            fprintf(stderr, "FALSE\n");
      }
#endif /* X_DEBUG */
